man/if1.1
.\" -*- NROFF -*-
.\" 
.\" if1.1
.\" 
.\" Author: Kevin Rosenberg <kevin@ctsim.org>
.\" 
.\" $Id: if1.1,v 1.5 2002/05/03 10:45:52 kevin Exp $
.\" 
.TH "if1" "1" "" "Kevin M. Rosenberg" "Engineering"
.SH "NAME"
if1 \- Single IF (Image File) Transformation
.SH "SYNOPSIS"
.B if1 iffile outfile [OPTIONS]
.SH "DESCRIPTION "
\fIif1\fP transforms a single IF file.
.SH "OPTIONS"
\fIif1\fP accepts the following options:
.TP 16 
.B \-\-invert
Inverts (negates) the image.
.TP 16
.B \-\-log
Natural logrithm of image.
.TP 16
.B \-\-exp
Natural exponential of image.
.TP 16
.B \-\-debug
Set debug mode.
.TP 16
.B \-\-verbose
Set verbose mode.
.TP 16
.B \-\-version
Print version.
.TP 16
.B \-\-help
Print terse help summary.
.SH "AUTHORS"
Kevin Rosenberg, M.D. <kevin@ctsim.org>
.SH "HISTORY"
CTSim was begun in 1983 using MS\-DOS and an EGA display adapter. In
1999 it was ported to GNU/Linux and later ported to Microsoft Windows.
.SH "SEE ALSO"
.BR ctsim (1)
.BR ctsimtext (1)
.BR if2 (1)
.BR ifexport (1)
.BR ifinfo (1)
.BR linogram (1)
.BR phm2helix (1)
.BR phm2if (1)
.BR phm2pj (1)
.BR pj2if (1)
.BR pjHinterp (1)
.BR pjinfo (1)
.BR pjrec (1)


NEWS
Version 4.2 Released Mar 2003

* Direct Fourier Reconstructions now supported

Version 4.0 Released Jan 2003

* Port to wxWindows 2.4

* Port to Mac OS X


Version 3.5 Released May 2002

* New reconstruction technique: Interpolate divergent beams to
  parallel beams for faster reconstructions.

* Multithreading for running length processes in background. Also
  takes advantage of multiple-CPU's in SMP systems for speed improvements
  with parallel processing.

* Region of interest can be specified for reconstruction zooming.

* Import PNG, PPM, and PGM images

* DICOM file import/export

* Clipboard for cutting and pasting images

* Can read and reconstruct SIEMENS AR.STAR DICOM raw data files

* Scattergram of T-Theta sampling. Useful for understanding divergent beam
  scanning.

* Startup-tips to help new users learn features.

* Ability to place detector at any arbitray position

* Histogram plotting of projection files


Version 3.0 New Features

* Greatly improved dialog boxes

* Online and print manuals with context-sensitive help!

* 3-d rotating views of image files

* Creation of arbitrary filter images

* Cubic interpolation for reconstructions

* All features of command line tools are now in graphical ctsim program!

* Complex-valued image files now supported, lots of image math
 functions added including Fourier transformations

* Visual and statistical image comparision functions

* Plotting of row and column data of single and comparison images

* Histogram Plotting of Image files.

* Conversion of projections to polar images

* Improved animation graphics

* New Microsoft Windows compatible self-installer with explorer extensions 
to CTSim files



Version 2.5 of CTSim has been released!

New Features of CTSim version 2.5:

- Now compiles with Microsoft Visual C++.

- First Microsoft Windows graphical user interface

- Bug fixes and modest documentation and program improvements over version 2.0



New Features of CTSim version 2.0:

- Entire code-base cleaned up, re-written, and converted from C to C++.

- Graphical user interface now included! Animation of projection and
  reconstruction processes. Displays projection and image files. Uses
  platform-independent wxWindows library so can run on UNIX and 
  Microsoft Windows platforms.

- Integrated G2 library for graphical display from command line tools
  (UNIX only)

- Added frequency-based filtering. Can provide significant speed-up
  in reconstructions compared to convolution-based filtering.

- Added highly-optimized backprojection method (idiff3)

- New command-line tools added for image statistics and comparisons with
  both text and graphical outputs.


tools/if1.cpp
/*****************************************************************************
** FILE IDENTIFICATION
**
**   Name:          if1.cpp
**   Purpose:       Manipulate a single image file
**   Programmer:    Kevin Rosenberg
**   Date Started:  Aug 1984
**
**  This is part of the CTSim program
**  Copyright (C) 1983-2009 Kevin Rosenberg
**
**  This program is free software; you can redistribute it and/or modify
**  it under the terms of the GNU General Public License (version 2) as
**  published by the Free Software Foundation.
**
**  This program is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
**  GNU General Public License for more details.
**
**  You should have received a copy of the GNU General Public License
**  along with this program; if not, write to the Free Software
**  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
******************************************************************************/

/* FILE
 *   if1.c             Filter a single IF file
 */

#include "ct.h"

enum {O_LOG, O_EXP, O_SQRT, O_SQR, O_INVERT, O_VERBOSE, O_HELP, O_VERSION};

static struct option my_options[] =
{
  {"invert", 0, 0, O_INVERT},
  {"verbose", 0, 0, O_VERBOSE},
  {"log", 0, 0, O_LOG},
  {"exp", 0, 0, O_EXP},
  {"sqr", 0, 0, O_SQR},
  {"sqrt", 0, 0, O_SQRT},
  {"help", 0, 0, O_HELP},
  {"version", 0, 0, O_VERSION},
  {0, 0, 0, 0}
};

static const char* g_szIdStr = "$Id$";

void
if1_usage (const char *program)
{
  std::cout << "usage: " << fileBasename(program) << " infile outfile [OPTIONS]" << std::endl;
  std::cout << "Generate a IF file from a IF file" << std::endl;
  std::cout << std::endl;
  std::cout << "     --invert   Invert image" << std::endl;
  std::cout << "     --log      Natural logrithm of image" << std::endl;
  std::cout << "     --exp      Natural exponential of image" << std::endl;
  std::cout << "     --sqr      Square of image" << std::endl;
  std::cout << "     --sqrt     Square root of image" << std::endl;
  std::cout << "     --verbose  Verbose mode" << std::endl;
  std::cout << "     --version  Print version" << std::endl;
  std::cout << "     --help     Print this help message" << std::endl;
}

int
if1_main (int argc, char *const argv[])
{
  char *in_file;
  char *out_file;
  int opt_verbose = 0;
  int opt_invert = 0;
  int opt_log = 0;
  int opt_exp = 0;
  int opt_sqr = 0;
  int opt_sqrt = 0;

  while (1)
    {
      int c = getopt_long (argc, argv, "", my_options, NULL);

      if (c == -1)
        break;

      switch (c)
        {
        case O_INVERT:
          opt_invert = 1;
          break;
        case O_LOG:
          opt_log = 1;
          break;
        case O_SQR:
          opt_sqr = 1;
          break;
        case O_SQRT:
          opt_sqrt = 1;
          break;
        case O_EXP:
          opt_exp = 1;
          break;
        case O_VERBOSE:
          opt_verbose = 1;
          break;
        case O_VERSION:
#ifdef VERSION
          std::cout << "Version " << VERSION << std::endl << g_szIdStr << std::endl;
#else
          std::cout << "Unknown version number" << std::endl;
#endif
          return (0);
        case O_HELP:
        case '?':
          if1_usage(argv[0]);
          return (0);
        default:
          if1_usage(argv[0]);
          return (1);
        }
    }

  if (optind + 2 != argc)
    {
      if1_usage(argv[0]);
      return (1);
    }

  in_file = argv[optind];
  out_file = argv[optind + 1];

  std::string histString;

  if (opt_invert || opt_log || opt_exp || opt_sqr || opt_sqrt) {
    ImageFile im_in;
    im_in.fileRead (in_file);
    int nx = im_in.nx();
    int ny = im_in.ny();
    ImageFile im_out (nx, ny);

    if (opt_invert) {
      im_in.invertPixelValues (im_out);
      histString = "Invert transformation";
    }
    if (opt_log) {
      im_in.log (im_out);
      histString = "Logrithmic transformation";
    }
    if (opt_exp) {
      im_in.exp (im_out);
      histString = "Exponential transformation";
    }
    if (opt_sqr) {
      im_in.square (im_out);
      histString = "Square transformation";
    }
    if (opt_sqrt) {
      im_in.sqrt (im_out);
      histString = "Square root transformation";
    }

    im_out.labelsCopy (im_in);
    im_out.labelAdd (Array2dFileLabel::L_HISTORY, histString.c_str());
    im_out.fileWrite (out_file);
  }

  return (0);
}

#ifndef NO_MAIN
int
main (int argc, char *const argv[])
{
  int retval = 1;

  try {
    retval = if1_main(argc, argv);
  } catch (exception e) {
    std::cerr << "Exception: " << e.what() << std::endl;
  } catch (...) {
    std::cerr << "Unknown exception" << std::endl;
  }

  return (retval);
}
#endif

